porch/controllers/remoterootsyncsets/pkg/remoteclient/getremoteclient.go

```diff
@@ -18,47 +18,95 @@ import (
 	"context"
 	"encoding/base64"
 	"fmt"
-	"os/exec"
-	"regexp"
+	"os"
 
 	container "cloud.google.com/go/container/apiv1"
-	api "github.com/GoogleContainerTools/kpt/porch/controllers/remoterootsyncsets/api/v1alpha1"
+	"github.com/GoogleContainerTools/kpt/porch/pkg/googleurl"
 	"golang.org/x/oauth2"
 	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
 	containerpb "google.golang.org/genproto/googleapis/container/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/discovery"
+	memory "k8s.io/client-go/discovery/cached"
+	"k8s.io/client-go/dynamic"
 	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
 	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// GetCCRESTConfig builds a rest.Config for accessing the config controller cluster,
+const (
+	containerClusterKind       = "ContainerCluster"
+	containerClusterApiVersion = "container.cnrm.cloud.google.com/v1beta1"
+
+	configControllerKind       = "ConfigControllerInstance"
+	configControllerApiVersion = "configcontroller.cnrm.cloud.google.com/v1beta1"
+)
+
+type RemoteClientGetter struct {
```

Comment:
We should figure out a place to put code that is shared between the controllers. It seems a bit strange to have this code under the remoterootsyncset controller when it is pretty generic and also used by the RootSyncSet controller.

Reply:
FWIW, in the rollouts branch, I refactored this into a separate package `clusterstore` to provides a way to list clusters, get clients (dynamic and client.Client) given a cluster. Eventually the plan was clusterstore could be an abstraction that can work with KCC clusters, hub clusters etc.

https://github.com/GoogleContainerTools/kpt/tree/feature-rollouts/rollouts/pkg/clusterstore

Reply:
Yes, agreed.  I was planning to do that as a follow-on.  I was even wondering if we should keep porch as a separate module, as it's a (minor) barrier to sharing with kpt to have to bump the versions.  Although that's less relevant for this logic, which is very server-side!

I like pkg/clusterstore :-)

Reply:
Yeah, I'm fine with this in a follow-up. I think we want to keep the modules separate and instead look into how we can make it easier to work with. At some point we will need to look into separate some of these components, for example splitting out some of the controllers that perhaps isn't tied to Porch.